File: uLSIF_code/demo_uLSIF.m
% demo_uLSIF.m
%
% (c) Masashi Sugiyama, Department of Compter Science, Tokyo Institute of Technology, Japan.
%     sugi@cs.titech.ac.jp,     http://sugiyama-www.cs.titech.ac.jp/~sugi/software/uLSIF/
clc
clear all

rand('state',0);
randn('state',0);

%%%%%%%%%%%%%%%%%%%%%%%%% Generating data
d=1;

dataset=1;
switch dataset
 case 1
  n_de=100;
  n_nu=100;
  mu_de=1;
  mu_nu=1;
  %p_de = 0.5;
  %p_nu = 0.1;
  sigma_de=1/2;
  sigma_nu=1/8;
  legend_position='northeast';
 case 2
  n_de=200;
  n_nu=1000;
  mu_de=1;
  mu_nu=2;
%   p_de = 0.6;
%   p_nu = 0.3;
  sigma_de=1/2;
  sigma_nu=1/4;
  legend_position='northwest';
end
% x_de = binornd(mu_de, p_de, 1, n_de);
% x_nu = binornd(mu_nu, p_nu, 1, n_nu);
x_de=mu_de+sigma_de*randn(d,n_de);
x_nu=mu_nu+sigma_nu*randn(d,n_nu);

% x_disp = linspace(0, max(mu_de, mu_nu) + 2, 100);
x_disp=linspace(-0.5,3,100);
% p_de_x_disp = binopdf(x_disp, mu_de, p_de);
% p_nu_x_disp = binopdf(x_disp, mu_nu, p_nu);
p_de_x_disp=pdf_Gaussian(x_disp,mu_de,sigma_de);
p_nu_x_disp=pdf_Gaussian(x_disp,mu_nu,sigma_nu);
w_x_disp=p_nu_x_disp./p_de_x_disp;

% p_de_x_de = binopdf(x_de, mu_de, p_de);
% p_nu_x_de = binopdf(x_de, mu_nu, p_nu);
p_de_x_de=pdf_Gaussian(x_de,mu_de,sigma_de);
p_nu_x_de=pdf_Gaussian(x_de,mu_nu,sigma_nu);
w_x_de=p_nu_x_de./p_de_x_de;

%%%%%%%%%%%%%%%%%%%%%%%%% Estimating density ratio
%[wh_x_de,wh_x_disp]=uLSIF(x_de,x_nu,x_disp);
[wh_x_de,wh_x_disp]=uLSIF(x_de,x_nu,x_disp,[],[],[],0);



figure(1)
clf
hold on
set(gca,'FontName','Helvetica')
set(gca,'FontSize',12)
plot(x_disp,p_de_x_disp,'b-','LineWidth',2)
plot(x_disp,p_nu_x_disp,'k-','LineWidth',2)
legend({'p_{de}(x)','p_{nu}(x)'},'Location',legend_position)
xlabel('x')
set(gcf,'PaperUnits','centimeters');
set(gcf,'PaperPosition',[0 0 12 9]);
print('-depsc',sprintf('density%g',dataset))

figure(2)
clf
hold on
set(gca,'FontName','Helvetica')
set(gca,'FontSize',12)
plot(x_disp,w_x_disp,'r-','LineWidth',2)
plot(x_disp,wh_x_disp,'g-','LineWidth',2)
plot(x_de,wh_x_de,'bo','LineWidth',1,'MarkerSize',8)
legend({'w(x)','w-hat(x)','w-hat(x^{de})'},'Location',legend_position)
xlabel('x')
set(gcf,'PaperUnits','centimeters');
set(gcf,'PaperPosition',[0 0 12 9]);
print('-depsc',sprintf('importance%g',dataset))

File: uLSIF_code/demo_uLSIF.py
import numpy as np
import uLSIF as ul
import pdf_Gaussian as pdfG
import matplotlib.pyplot as plt
import pandas as pd
np.random.seed(0)

d = 1
dataset = 1
if dataset == 1:
    n_de = 100
    n_nu = 100
    mu_de = 1
    mu_nu = 1
    sigma_de = 1 / 2
    sigma_nu = 1 / 8
    legend_position = 'upper right'
elif dataset == 2:
    n_de = 200
    n_nu = 1000
    mu_de = 1
    mu_nu = 2
    sigma_de = 1 / 2
    sigma_nu = 1 / 4
    legend_position = 'upper left'

# Generate data for domain adaptation
x_de = np.random.normal(loc=mu_de, scale=sigma_de, size=(d, n_de))
x_nu = np.random.normal(loc=mu_nu, scale=sigma_nu, size=(d, n_nu))

x_disp = np.array([np.linspace(-0.5,3,100)])

p_de_x_disp = pdfG.pdf_Gaussian(x_disp,mu_de,sigma_de)
p_nu_x_disp = pdfG.pdf_Gaussian(x_disp,mu_nu,sigma_nu)
w_x_disp=p_nu_x_disp/p_de_x_disp

p_de_x_de = pdfG.pdf_Gaussian(x_de,mu_de,sigma_de)
p_nu_x_de = pdfG.pdf_Gaussian(x_de,mu_nu,sigma_nu)
w_x_de=p_nu_x_de/p_de_x_de

wh_x_de,wh_x_disp =ul.uLSIF(x_de,x_nu,x_disp,np.logspace(-3, 1, 9),np.logspace(-3, 1, 9),100,5)


plt.figure(1)
plt.plot(x_disp[0], p_de_x_disp, "b")    
plt.plot(x_disp[0], p_nu_x_disp, "r")    
plt.legend(loc=legend_position,labels=['p_de(x)','p_nu(x)'])

plt.figure(2)
plt.plot(x_disp[0], w_x_disp, "b")    
plt.plot(x_disp[0], wh_x_disp, "r")  
plt.scatter(x_de[0], wh_x_de)
plt.legend(loc=legend_position,labels=['w(x)','w_hat(x)','w-hat(x^de)'])

plt.show() 

